### DISCUSSION TITLE
Remove aspects

### DISCUSSION THREAD

**Discussion by josefd8** [veepee-oss/gingerspec#107] [Ideas]
Hello,

I would like to start a process to remove (as many) aspects as possible. The idea would be to refactor them using other techniques to achieve the same functionality (without breaking the QAs projects) and at the same time, remove aspects that I consider unnecessary or that are rarely used.

I feel that using aspects (sometimes) bring more drawbacks than benefits for the following reasons:

* They are complicated to debug and if an exception is thrown within an aspect, it is complicated to locate exactly the source of the problem.
* All aspects are executed even when not required.
* Sometimes, interrupting the normal execution of scenarios using aspects causes unintended consequences, for example, certain information not being correctly displayed in the reports
* And for me, the biggest problem with aspects is they rely too heavily on cucumber internal implementation instead of using the public APIs that the framework provides (and that are much more stable). For example, if I wanted to update the cucumber dependency, this may cause several of the aspects to break, since they rely on methods within cucumber that were not meant to be public and that need to be accessed using reflection. So with every change in the cucumber dependency comes the risk of a painful refactor process of several aspects.

- [@ignore aspect](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#ignore-tag)
   This one could easily be implemented in a hook. Scenarios will be printed in the console in light blue color to indicate they were skipped and they will appear as "skipped" in the reports

- [@runOnEnv and @skipOnEnv aspects](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#runonenv-tag)
  This could also be implemented in a hook just like the @ignore aspect

- [@loop aspect](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#loop-tag)
  I would remove this aspect from Gingerspec. The same functionality can be achieved using a scenario outline (in fact, what this aspect does is transform the scenario into a scenario outline). The only good thing about this one is that the scenario outline is constructed dynamically using a variable, so I can see it being useful in some particular cases, but I really don't know anyone making use of it.

- [@background tag](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#background-tag) 
  I would also remove this one, haven't seen anybody using it. Although I admit it might come in handy in some situations. The problem with this one is that it only accepts maven variables (-DVAR=VALUE), it doesn't work with variables created during the scenario, so it doesn't really work as an IF statement.

- [@log aspect](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#background-tag)
  I could probably find a way of implementing this one using a beforeStep hook

- [@include aspect](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-tags#background-tag)
  This is a complicated one. It might be useful in some situations but I haven't seen people using it that much and I feel it creates scenarios that are harder to read and maintain. I would also remove this one

- [replacement aspect](https://github.com/veepee-oss/gingerspec/wiki/Gherkin-variables)
  This is the only aspect that I wouldn't dare to touch for the moment and that cannot (at least for now) be rewritten in any other way that is not an aspect. I would leave this one as it is but if I find a way of implementing this without an aspect, I would do it

I would like to hear your opinion @oleksandr0001  on this. Maybe what is not-so-useful for me, could be really useful for someone else

**Comment by oleksandr0001** [comment]
It's a great idea. But does it affect the scenario o feature view? (for example, after the change how do you use the log o ignore tag? inside the feature file the same way?)
